ros2_socketcan/src/socket_can_receiver_node.cpp

```diff
@@ -108,37 +135,76 @@ LNI::CallbackReturn SocketCanReceiverNode::on_shutdown(const lc::State & state)
 void SocketCanReceiverNode::receive()
 {
   CanId receive_id{};
-  can_msgs::msg::Frame frame_msg(rosidl_runtime_cpp::MessageInitialization::ZERO);
-  frame_msg.header.frame_id = "can";
-
-  while (rclcpp::ok()) {
-    if (this->get_current_state().id() != State::PRIMARY_STATE_ACTIVE) {
-      std::this_thread::sleep_for(100ms);
-      continue;
-    }
 
-    try {
-      receive_id = receiver_->receive(frame_msg.data.data(), interval_ns_);
-    } catch (const std::exception & ex) {
-      RCLCPP_WARN_THROTTLE(
-        this->get_logger(), *this->get_clock(), 1000,
-        "Error receiving CAN message: %s - %s",
-        interface_.c_str(), ex.what());
-      continue;
+  if (!enable_fd_) {
+    can_msgs::msg::Frame frame_msg(rosidl_runtime_cpp::MessageInitialization::ZERO);
+    frame_msg.header.frame_id = "can";
+
+    while (rclcpp::ok()) {
+      if (this->get_current_state().id() != State::PRIMARY_STATE_ACTIVE) {
+        std::this_thread::sleep_for(100ms);
+        continue;
+      }
+
+      try {
+        receive_id = receiver_->receive(frame_msg.data.data(), interval_ns_);
+      } catch (const std::exception & ex) {
+        RCLCPP_WARN_THROTTLE(
+          this->get_logger(), *this->get_clock(), 1000,
+          "Error receiving CAN message: %s - %s",
+          interface_.c_str(), ex.what());
+        continue;
+      }
+
+      if (use_bus_time_) {
+        frame_msg.header.stamp =
+          rclcpp::Time(static_cast<int64_t>(receive_id.get_bus_time() * 1000U));
+      } else {
+        frame_msg.header.stamp = this->now();
+      }
+
+      frame_msg.id = receive_id.identifier();
+      frame_msg.is_rtr = (receive_id.frame_type() == FrameType::REMOTE);
+      frame_msg.is_extended = receive_id.is_extended();
+      frame_msg.is_error = (receive_id.frame_type() == FrameType::ERROR);
+      frame_msg.dlc = receive_id.length();
+      frames_pub_->publish(std::move(frame_msg));
     }
-
-    if (use_bus_time_) {
-      frame_msg.header.stamp =
-        rclcpp::Time(static_cast<int64_t>(receive_id.get_bus_time() * 1000U));
-    } else {
-      frame_msg.header.stamp = this->now();
+  } else {
+    ros2_socketcan_msgs::msg::FdFrame fd_frame_msg(rosidl_runtime_cpp::MessageInitialization::ZERO);
+    fd_frame_msg.header.frame_id = "can";
+    fd_frame_msg.data.resize(64);
+
+    while (rclcpp::ok()) {
+      if (this->get_current_state().id() != State::PRIMARY_STATE_ACTIVE) {
+        std::this_thread::sleep_for(100ms);
+        continue;
+      }
+      try {
+        receive_id = receiver_->receive_fd(fd_frame_msg.data.data<void>(), interval_ns_);
+      } catch (const std::exception & ex) {
+        RCLCPP_WARN_THROTTLE(
+          this->get_logger(), *this->get_clock(), 1000,
+          "Error receiving CAN FD message: %s - %s",
+          interface_.c_str(), ex.what());
+        continue;
+      }
+
+      fd_frame_msg.data.resize(receive_id.length());
```

Comment:
Hello,
I am testing this package and I found the following issue:

The fd_frame_msg.data is resized at the beginning to 64  and passed to receive_fd() to be filled (memcpy)
Then the fd_frame_msg.data is resized to the actual payload length before publish()

In the next loop the fd_frame_msg.data must be resized to 64

Reply:
Thanks for catching this! It has been corrected.